### PR TITLE
fix query adjacent by filtering conditions with limit

### LINKED ISSUE #1052: limit语句执行时机有误，导致查询结果不准确。
### Expected behavior 期望表现
形如语句`g.V('1:837401253').bothE('know').has('act_id', '50069').limit(100)`

在没有limit条件下可以查询出500条数据，期望返回100条数据。

### Actual behavior 实际表现
实际返回5条数据，

### Steps to reproduce the problem 复现步骤

似乎先执行了limit和bothE，再执行has语句，导致查询结果变少。

### PR DESCRIPTION
**Pull request by javeme**
fix #1052

Change-Id: I11075c56e89a57cbf8e599a2e5fa6c0122663f6d

### CODE REVIEW

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java [hunk 1]
```diff
@@ -1932,6 +1932,430 @@ public void testQueryEdgesWithLimitOnMultiLevel() {
...
+
+        int txCap = this.superNodeSize();
+        for (int i = 0; i < txCap; i++) {
+            Vertex james = graph.addVertex(T.label, "author", "id", 1,
+                                           "name", "James Gosling", "age", 62,
+                                           "lived", "Canadian");
+            Vertex java = graph.addVertex(T.label, "book", "name", "java-" + i);
+            james.addEdge("authored", java, "score", i % 2);
+            james.addEdge("write", java, "time", "2020-6-" + i);
+            if (i % TX_BATCH == 0) {
```
Comment (houzhizhen) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java:2347]: first batch only loop 1 time

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java [hunk 2]
```diff
@@ -1932,6 +1932,430 @@ public void testQueryEdgesWithLimitOnMultiLevel() {
...
+        Assert.assertEquals(1, vertices.size());
+    }
+
+    @Test
+    public void testQueryEdgesWithLimitOnSuperVertexAndFilterProp() {
+        HugeGraph graph = graph();
+
+        int txCap = this.superNodeSize();
+        for (int i = 0; i < txCap; i++) {
+            Vertex james = graph.addVertex(T.label, "author", "id", 1,
```
Comment (houzhizhen) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java:2341]: add duplicate vertex in every loop?

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java [hunk 3]
```diff
@@ -1932,6 +1932,430 @@ public void testQueryEdgesWithLimitOnMultiLevel() {
...
+
+        vertices = graph.traversal().V(jeff)
+                        .both().has("age", 61)
+                        .limit(3).toList();
+        Assert.assertEquals(2, vertices.size());
+
+        vertices = graph.traversal().V(jeff)
+                        .both().has("age", 61)
+                        .limit(1).toList();
+        Assert.assertEquals(1, vertices.size());
```
Comment (houzhizhen) [inline hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java:2332]: what about limit 0?

**Comment by javeme** [comment]
ci error with non rocksdb/hbase backends:
```
[ERROR] Tests run: 672, Failures: 1, Errors: 0, Skipped: 62, Time elapsed: 239.451 s <<< FAILURE! - in com.baidu.hugegraph.core.CoreTestSuite
[ERROR] testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 0.037 s  <<< FAILURE!
java.lang.AssertionError: expected:<11> but was:<9>
	at com.baidu.hugegraph.core.EdgeCoreTest.testQueryEdgesAdjacentVerticesWithLimitAndFilterProp(EdgeCoreTest.java:2222)

```

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java (hunk 1 above)
Comment (javeme) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java:2347]: not affect logic

**Comment by javeme** [comment]
`g.V().out("write", "look").limit(12)` with cassandra order:
```
e[S206:Louise>611>2020-6-0>S209:java-0][206:Louise-look->209:java-0]
e[S206:Louise>611>2020-6-18>S209:java-0][206:Louise-look->209:java-0]

e[S207:11>610>2020-6-0>S209:java-0][207:11-write->209:java-0]
e[S207:11>610>2020-6-1>S209:java-1][207:11-write->209:java-1]
e[S207:11>610>2020-6-10>S209:java-10][207:11-write->209:java-10]
e[S207:11>610>2020-6-11>S209:java-11][207:11-write->209:java-11]
e[S207:11>610>2020-6-12>S209:java-12][207:11-write->209:java-12]
e[S207:11>610>2020-6-13>S209:java-13][207:11-write->209:java-13]
e[S207:11>610>2020-6-14>S209:java-14][207:11-write->209:java-14]
e[S207:11>610>2020-6-15>S209:java-15][207:11-write->209:java-15]
e[S207:11>610>2020-6-16>S209:java-16][207:11-write->209:java-16]
e[S207:11>610>2020-6-17>S209:java-17][207:11-write->209:java-17]
e[S207:11>610>2020-6-18>S209:java-18][207:11-write->209:java-18]
```

**Comment by javeme** [comment]
pgsql error:
```
[ERROR] Tests run: 673, Failures: 0, Errors: 1, Skipped: 21, Time elapsed: 289.197 s <<< FAILURE! - in com.baidu.hugegraph.core.CoreTestSuite
[ERROR] testAddEdgeWithInvalidSortkey(com.baidu.hugegraph.core.EdgeCoreTest)  Time elapsed: 0.14 s  <<< ERROR!
com.baidu.hugegraph.backend.BackendException: Failed to commit
	at com.baidu.hugegraph.core.EdgeCoreTest.testAddEdgeWithInvalidSortkey(EdgeCoreTest.java:512)
Caused by: java.sql.BatchUpdateException: Batch entry 0 INSERT INTO g_oe (PROPERTIES, OTHER_VERTEX, LABEL, DIRECTION, SORT_VALUES, OWNER_VERTEX) VALUES ('{"5033":"2017-5-27\u0000"}', 'S2237:Test-Book-1', 262, -126, '2017-5-272017-5-27', 'S2235:11') ON CONFLICT (OWNER_VERTEX, DIRECTION, LABEL, SORT_VALUES, OTHER_VERTEX) DO UPDATE SET PROPERTIES = '{"5033":"2017-5-27\u0000"}', OTHER_VERTEX = 'S2237:Test-Book-1', LABEL = 262, DIRECTION = -126, SORT_VALUES = '2017-5-272017-5-27', OWNER_VERTEX = 'S2235:11' was aborted: ERROR: invalid byte sequence for encoding "UTF8": 0x00  Call getNextException to see other errors in the batch.
	at com.baidu.hugegraph.core.EdgeCoreTest.testAddEdgeWithInvalidSortkey(EdgeCoreTest.java:512)
Caused by: org.postgresql.util.PSQLException: ERROR: invalid byte sequence for encoding "UTF8": 0x00
	at com.baidu.hugegraph.core.EdgeCoreTest.testAddEdgeWithInvalidSortkey(EdgeCoreTest.java:512)
```

File: hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java (hunk 3 above)
Comment (houzhizhen) [reply hugegraph-test/src/main/java/com/baidu/hugegraph/core/EdgeCoreTest.java:2487]: It's better to add a limit(0) boundary test.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java [hunk 4]
```diff
@@ -332,14 +333,27 @@ public String readString() {
...
+    public BytesBuffer writeStringWithEnding(String value) {
+        if (!value.isEmpty()) {
+            byte[] bytes = StringEncoding.encode(value);
+            /*
+             * assert '0x00'/'0xFF' not exist in string index id
+             * NOTE:
+             *   0x00 is NULL in UTF8(or ASCII) bytes
+             *   0xFF is not a valid byte in UTF8 bytes
+             */
+            assert !Bytes.contains(bytes, STRING_ENDING_BYTE_FF) :
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java:345]: if not use 0xff at all, prefer delete it and no need to assert

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 5]
```diff
@@ -1066,6 +1045,9 @@ private static ConditionQuery constructQuery(ConditionQuery query,
                              indexType, indexFields);
                 Object fieldValue = query.userpropValue(indexFields.get(0));
                 assert fieldValue instanceof String;
+                // Escape empty String to INDEX_EMPTY_SYM
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:1048]: INDEX_EMPTY_SYM -> INDEX_SYM_EMPTY

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 6]
```diff
@@ -1074,11 +1056,9 @@ private static ConditionQuery constructQuery(ConditionQuery query,
             case SECONDARY:
                 List<Id> joinedKeys = indexFields.subList(0, queryKeys.size());
                 String joinedValues = query.userpropValuesString(joinedKeys);
+                // Escape empty String to INDEX_EMPTY_SYM
```
Comment (zhoney) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:1059]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java (hunk 4 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/serializer/BytesBuffer.java:345]: to preserve the byte for future usages

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java [hunk 7]
```diff
@@ -1189,7 +1195,9 @@ public static boolean matchEdgeSortKeys(ConditionQuery query,
         for (int i = sortKeys.size(); i > 0; i--) {
             List<Id> subFields = sortKeys.subList(0, i);
             if (queryKeys.containsAll(subFields)) {
-                return true;
+                if (queryKeys.size() == subFields.size() || !matchAll) {
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java:1198]: add a comment later

File: hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java [hunk 8]
```diff
@@ -124,9 +124,15 @@ public HugeVertexStep(final VertexStep<E> originVertexStep) {
         // Query by sort-keys
         if (withEdgeCond && edgeLabels.length > 0) {
             TraversalUtil.fillConditionQuery(conditions, query, graph);
-            if (!GraphTransaction.matchEdgeSortKeys(query, graph)) {
+            if (!GraphTransaction.matchEdgeSortKeys(query, false, graph)) {
```
Comment (Linary) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/optimize/HugeVertexStep.java:127]: suggest rename to partialMatchEdgeSortKeys

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java (hunk 7 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphTransaction.java:1205]: done

**Comment by javeme** [comment]
memory ci:
```
[ERROR] Tests run: 673, Failures: 1, Errors: 5, Skipped: 62, Time elapsed: 442.391 s <<< FAILURE! - in com.baidu.hugegraph.core.CoreTestSuite
[ERROR] testTask(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 2.173 s  <<< FAILURE!
java.lang.AssertionError: Bad exception type java.lang.IllegalArgumentException(expected com.baidu.hugegraph.HugeException)
	at com.baidu.hugegraph.core.TaskCoreTest.testTask(TaskCoreTest.java:85)
[ERROR] testGremlinJobWithError(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.009 s  <<< ERROR!
java.lang.IllegalArgumentException: Can't delete incomplete task '88888' in status RUNNING, Please try to cancel the task first
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
[ERROR] testTaskWithFailure(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.006 s  <<< ERROR!
java.lang.IllegalArgumentException: Can't delete incomplete task '88888' in status RUNNING, Please try to cancel the task first
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
[ERROR] testEphemeralJob(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.004 s  <<< ERROR!
java.lang.IllegalArgumentException: Can't delete incomplete task '88888' in status RUNNING, Please try to cancel the task first
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
[ERROR] testGremlinJobWithFailure(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.002 s  <<< ERROR!
java.lang.IllegalArgumentException: Can't delete incomplete task '88888' in status RUNNING, Please try to cancel the task first
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
[ERROR] testGremlinJob(com.baidu.hugegraph.core.TaskCoreTest)  Time elapsed: 0.004 s  <<< ERROR!
java.lang.IllegalArgumentException: Can't delete incomplete task '88888' in status RUNNING, Please try to cancel the task first
	at com.baidu.hugegraph.core.TaskCoreTest.setup(TaskCoreTest.java:64)
```

**Pull request merged**